### PR TITLE
Fix memory leak

### LINKED ISSUE #6: Potential memory leak
Hi,
as always - thanks for working on and maintaining the async suite of gems 🙏 

I *think* I found a memory leak in this gem. Here's a repro example using async redis

```ruby
require "async"
require "async/redis"

Async do |task|
	c = Async::Redis::Client.new

	10_000.times do |i|
		c.ping "foo#{i}"
	end

	available = c.instance_variable_get("@pool").instance_variable_get("@available")
	puts "available size #{available.size}, unique #{available.uniq.size}"
end
```

Outputs: `available size 10000, unique 1`.

I discovered this after `Async::Pool::Controller` `@available` array grew to ~200Mb after 15 hours in a process that used `async-redis`.

I think the problem happens because a resource from `@available` array is `acquire`d here without being removed from the array:

https://github.com/socketry/async-pool/blob/c5cdfe56db91680c561ffb070a6815cdafe06717/lib/async/pool/controller.rb#L214-L217

Later, a duplicate is added to `@available` array again:

https://github.com/socketry/async-pool/blob/c5cdfe56db91680c561ffb070a6815cdafe06717/lib/async/pool/controller.rb#L173-L180

I'll try to submit a PR that fixes this.

### PR DESCRIPTION
**Pull request by bruno-**
This PR fixes #6.

I'm not sure how to test this tho. I tried comparing `GC.start; ObjectSpace.memsize_of_all` before and after running a couple `#acquire + #release` but that didn't work. It's probably not OK to use `pool.instance_variable_get("@available")`?

## Types of Changes

- [x] Bug fix.
- [ ] New feature.
- [ ] Performance improvement.

## Testing

- [ ] I added new tests for my changes.
- [x] I ran all the tests locally.

### CODE REVIEW

File: lib/async/pool/controller.rb [hunk 1]
```diff
@@ -201,26 +202,26 @@ def create_resource
...
+					@available.add(resource) if resource.concurrency > 1
 				end
 				
 				return resource
 			end
 			
 			def available_resource
 				@guard.acquire do
-					while resource = @available.last
+					while resource = @available.first
```
Comment (bruno-) [inline lib/async/pool/controller.rb:213]: Not sure if this change matters?

File: lib/async/pool/controller.rb [hunk 2]
```diff
@@ -34,7 +35,7 @@ def self.wrap(**options, &block)
 			def initialize(constructor, limit: nil)
 				@resources = {}
 				
-				@available = []
+				@available = Set.new
```
Comment (bruno-) [inline lib/async/pool/controller.rb:38]: Do you think we should use `#compare_by_identity`?

If yes, I have a commit and tests for that

https://github.com/socketry/async-pool/compare/master...bruno-:eql-resources?expand=1#diff-3ac6d80b31ece477a4f1199d388a70dbe5bb7945de57bd2603f2bd8a6d9c53d3

**Comment by ioquatix** [comment]
Ah you are right, there is a potential bug.

File: lib/async/pool/controller.rb (hunk 2 above)
Comment (ioquatix) [reply lib/async/pool/controller.rb:38]: The problem is, `pop` no longer works correctly removing the last item, right?

Comment (bruno-) [reply lib/async/pool/controller.rb:38]: No, everything in this PR works just fine. Tests pass and issue #6 is fixed.

`#compare_by_identity` addresses another potential problem:

If a resource class has `#eql?` and `#hash` methods overridden (a common pattern when working with `Set`s), `@resources` hash and `@available` set may de-duplicate two different resource instances.

The problem is hypothetic, and I'm probably not explaining it well. We can ignore `#compare_by_identity` in order to fix the memory leak.

Comment (ioquatix) [reply lib/async/pool/controller.rb:38]: So `Set#first` and `Set#pop` work? as in `pop` pops the first item?

**Comment by ioquatix** [comment]
For some context, `@available` is just supposed to be an optimisation on `@resources` to find an available resource quickly. Using a set is acceptable but first I want to add some failing test cases to master to scope out the full extent of the problem.

File: lib/async/pool/controller.rb (hunk 2 above)
Comment (bruno-) [reply lib/async/pool/controller.rb:38]: Hmm, no, method `Set#pop` does not exist.
This PR changes removing resources from `@available` to work like this:

```ruby
resource = @available.first
# work
@available.delete(resource) # efficient operation O(1)
```

**Comment by bruno-** [comment]
Closing in favor of d545ad7969bf670ff5d1d78c45d2cc89b0269983

**Pull request closed without merging**